cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -2884,9 +2884,9 @@
 (rule 2 (x64_imul_imm $I16 src1 (i8_try_from_i32 src2))  (x64_imulw_rmi_sxb src1 src2))
 (rule 2 (x64_imul_imm $I32 src1 (i8_try_from_i32 src2))  (x64_imull_rmi_sxb src1 src2))
 (rule 2 (x64_imul_imm $I64 src1 (i8_try_from_i32 src2))  (x64_imulq_rmi_sxb src1 src2))
-(rule 1 (x64_imul_imm $I16 src1 (u16_try_from_i32 src2)) (x64_imulw_rmi src1 src2))
-(rule 1 (x64_imul_imm $I32 src1 (i32_as_u32 src2))       (x64_imull_rmi src1 src2))
-(rule 1 (x64_imul_imm $I64 src1 src2)                    (x64_imulq_rmi_sxl src1 src2))
+(rule 1 (x64_imul_imm $I16 src1 (i16_try_from_i32 src2)) (x64_imulw_rmi src1 (i16_as_u16 src2)))
```

Comment:
I'll admit this is testing the limits of my CLIF knowledge. This will fail when the `i32` value fits in a `u16`, but not an `i16`, meaning that we'll still panic during instruction selection from that. I don't know whether that's possible in CLIF though, or if `iconst` values are always sign-extended from their type.

Reply:
Oh, missed that, sorry -- unfortunately we've standardized on zero extension instead (it keeps the majority of cases simpler). Fortunately if this extractor returns `None` then the rule just doesn't match -- no runtime panic unless there's not another fallback? Let's (i) add a test case with a value in that range (say 40000) and (ii) verify there is...

Reply:
Ok I've added a few CLIF tests for this and it looks like things are actually working as intended. Looks like [when accessing the constant in ISLE](https://github.com/bytecodealliance/wasmtime/blob/24730416736955f370b5074f1874cea2dab2784d/cranelift/codegen/src/machinst/isle.rs#L222) it gets sign-extended based on the type of the constant itself, which is why this ended up working out. Yay!